service/history/replication/executable_task.go

```diff
@@ -355,35 +364,45 @@ func (e *ExecutableTaskImpl) DeleteWorkflow(
 }
 
 func (e *ExecutableTaskImpl) GetNamespaceInfo(
+	ctx context.Context,
 	namespaceID string,
 ) (string, bool, error) {
 	namespaceEntry, err := e.NamespaceCache.GetNamespaceByID(namespace.ID(namespaceID))
 	switch err.(type) {
 	case nil:
-		e.namespace.Store(namespaceEntry.Name())
-		shouldProcessTask := false
-	FilterLoop:
-		for _, targetCluster := range namespaceEntry.ClusterNames() {
-			if e.ClusterMetadata.GetCurrentClusterName() == targetCluster {
-				shouldProcessTask = true
-				break FilterLoop
-			}
-		}
-		return string(namespaceEntry.Name()), shouldProcessTask, nil
 	case *serviceerror.NamespaceNotFound:
-		return "", false, nil
+		if !e.ProcessToolBox.Config.EnableReplicationEagerRefreshNamespace() {
+			return "", false, nil
+		}
+
+		namespaceEntry, err = e.ProcessToolBox.EagerNamespaceRefresher.SyncNamespaceFromSourceCluster(ctx, namespace.ID(namespaceID), e.sourceClusterName)
+		if err != nil {
+			e.Logger.Info("Failed to SyncNamespaceFromSourceCluster", tag.Error(err))
+			return "", false, nil
+		}
 	default:
 		return "", false, err
 	}
+
+	e.namespace.Store(namespaceEntry.Name())
+	shouldProcessTask := false
+FilterLoop:
+	for _, targetCluster := range namespaceEntry.ClusterNames() {
+		if e.ClusterMetadata.GetCurrentClusterName() == targetCluster {
+			shouldProcessTask = true
+			break FilterLoop
+		}
+	}
+	return string(namespaceEntry.Name()), shouldProcessTask, nil
 }
 
 func newTaskContext(
-	namespaceName string,
+	namespaceId string,
 ) (context.Context, context.CancelFunc) {
 	ctx := headers.SetCallerInfo(
 		context.Background(),
 		headers.SystemPreemptableCallerInfo,
 	)
-	ctx = headers.SetCallerName(ctx, namespaceName)
+	ctx = headers.SetCallerName(ctx, namespaceId)
```

Comment:
Will this be a problem if we change to namespace id?

Reply:
I am seeing a mixed usage of namespaceId and namespaceName on current code. So I assume it should not be a problem.

Reply:
I think we should use name. Could you let me know where we use id to SetCallerName?

Reply:
above link should probably use namespace name, not ID